drivers/adc/adc_gecko.c

```diff
@@ -0,0 +1,416 @@
+#include <kernel.h>
+#include <init.h>
+#include <soc.h>
+#include <em_adc.h>
+#include <em_cmu.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER
+#include "adc_context.h"
+
+#define LOG_LEVEL CONFIG_ADC_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(adc_gecko);
+
+#define ADC_PREFIX cmuClock_ADC
+#define CLOCK_ADC(id) _CONCAT(ADC_PREFIX, id)
+
+/* Acquisition times */
+static const u16_t acq_time_single_tbl[10] = {
+	1, 2, 3, 4, 8, 16, 32, 64, 128, 256
+};
+
+/* Number of channels available. */
+#define GECKO_CHANNEL_COUNT	32
+
+struct adc_gecko_ch_cfg {
+	ADC_Res_TypeDef resolution;
+	ADC_Ref_TypeDef reference;
+	ADC_AcqTime_TypeDef acq_time;
+	ADC_PosSel_TypeDef input_positive;
+	bool initialized;
+};
+
+struct adc_gecko_data {
+	struct adc_context ctx;
+	struct device *dev;
+	u16_t *buffer;
+	u16_t *repeat_buffer;
+	u32_t channels;
+	u8_t channel_id;
+	struct adc_gecko_ch_cfg ch_cfg[GECKO_CHANNEL_COUNT];
+};
+
+struct adc_gecko_cfg {
+	ADC_TypeDef *base;
+	CMU_Clock_TypeDef clock;
+	u32_t prescaler;
+	void (*irq_cfg_func)(void);
+};
+
+#define DEV_DATA(dev) \
+	((struct adc_gecko_data *)(dev)->driver_data)
+#define DEV_CFG(dev) \
+	((struct adc_gecko_cfg *)(dev)->config->config_info)
+
+static int adc_gecko_check_buffer_size(const struct adc_sequence *sequence,
+					u8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(u16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_ERR("Provided buffer is too small (%u/%u)",
+			sequence->buffer_size, needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_gecko_check_resolution(const struct adc_sequence *sequence)
+{
+	int value = 0;
+
+	switch (sequence->resolution) {
+	case 6:
+		value = adcRes6Bit;
+		break;
+	case 8:
+		value = adcRes8Bit;
+		break;
+	case 12:
+		value = adcRes12Bit;
+		break;
+	default:
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	return value;
+}
+
+static int start_read(struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_gecko_data *data = DEV_DATA(dev);
+	u32_t channels;
+	u8_t channel_count;
+	u8_t index;
+	int res;
+
+	/* Check at least 1 channel is requested */
+	if (sequence->channels == 0) {
+		LOG_ERR("No channel requested");
+		return -EINVAL;
+	}
+
+	/* Check oversampling setting */
+	if (sequence->oversampling != 0) {
+		LOG_ERR("Oversampling is not supported");
+		return -ENOTSUP;
+	}
+
+	/* Check resolution setting */
+	res = adc_gecko_check_resolution(sequence);
+	if (res < 0) {
+		return -EINVAL;
+	}
+
+	/* Verify all requested channels are initialized and store resolution */
+	channels = sequence->channels;
+	channel_count = 0;
+	while (channels) {
+		index = find_lsb_set(channels) - 1;
+		if (!data->ch_cfg[index].initialized) {
+			LOG_ERR("Channel not initialized");
+			return -EINVAL;
+		}
+		data->ch_cfg[index].resolution = res;
```

Comment:
It's a bit waste of memory to store the resolution for each channel while in API it is specified for the sampling sequence. I'd suggest to store this setting in `data->resolution` instead.